### DISCUSSION TITLE
Local console changes apply immediately for cloud

### DISCUSSION THREAD

**Discussion by erquhart** [hasura/graphql-engine#8928] [Question] [answered]
I'm certain I'm missing something really obvious, just need help understanding what that is.

When I run the console locally against my cloud project, it both generates migrations in my local file system and immediately applies changes as if I were making them on the hosted console. How do I keep changes local and have them only apply when expected (in this case when I push to main)?

**Comment by meetzaveri** [comment, marked as answer]
Hey @erquhart 

> When I run the console locally against my cloud project

Can you elaborate this please ? Is it that you are running console locally against your DB hosted which is connected with your cloud project ? 

###  Suggestion:
You can source control your Hasura cloud project with your local console repository via git deploy feature. What this “git deploy” feature will do is that whatever metadata and migrations you push, it will apply to that branch (git branch which you’d specified in Hasura cloud project) and then your Hasura cloud console will run those migrations and metadata. Check out here about git deploy docs here - https://hasura.io/docs/latest/deployment/github-integration/ . (See the screenshot on how structure of directory looks like which will be captured by git deploy feature)

   <img width="284" alt="hasura_git_deploy_directory_structure" src="https://user-images.githubusercontent.com/21293048/188566841-35df384d-5d4d-4207-9492-e7555092c08a.png">


 Essentially in development mode, we keep a clone of the original DB and use it for development purpose. This way we do isolate DB from which is separate as that of cloud project. Having console running locally against your DB (which is also running locally), this way you can keep migrations away from polluting your production DB (the one connected with your cloud project) and store them locally.
 
   All of the migrations and metadata can be generated locally via CLI and can be stored in your repository( source control ). The control is in your hands to push needed ones. Check CLI docs to find out which CLI command suits your use case. https://hasura.io/docs/latest/hasura-cli/commands/hasura/

**Reply by erquhart** [reply]
This is exactly what I need, thank you!

Your guide to local development specifically says to use the same database connection string as your cloud database: https://hasura.io/docs/latest/projects/environments/#step-1-setup-local-development-of-hasura-with-docker

It sounds like you're saying that this is only used to make a development clone? 

The immediate problem is that, when I make a change using the cli console, two things happen:
- the change is present in my local migration files, visible in git diff
- the change is already applied to my cloud database - if I navigate to the cloud console (not local), the change is reflected there, which means the local console is altering my production database in realtime 

I'm using Git deploy as you mentioned, but how do I keep changes through the local console from applying to my production database?

Here's the workflow that I expected, based on my understanding of the cli and Git deployments:
1. run local database on docker
2. apply metadata and migrations from production db to local db
3. use the cli console to make changes to local db only
4. commit the resulting metadata/migrations file changes
5. *at this point my production database and cloud metadata have not been altered, all changes are local only*
6. push to remote branch connected to hasura git deployments
7. Hasura applies metadata and migrations to Hasura cloud and production database

Trying to understand what part of this I'm incorrect on. Thanks again for your help!

**Reply by meetzaveri** [reply]
Hey @erquhart 

Quoting one of the lines from your message
> Your guide to local development specifically says to use the same database connection string as your cloud database:

So are you referring to this line in that docs (see the below screenshot) ?
![captures_chrome-capture-2022-8-7](https://user-images.githubusercontent.com/21293048/188829666-e61da058-6ab8-4d06-8b2c-3a7ebeb24905.png)

If yes, then it does state that you should use same ENV variable "name" (key as in key:value pair) that you were using in cloud. It means that if `PG_DATABASE_URL` was the env var in cloud for storing database connection string, then you should use it as `PG_DATABASE_URL` in local too (for the database connection string, regardless of the value that may differ). 

It does not _explicitly_ state that you'll need to pass the **_same value_**(i.e. connection string) . The database connection string you pass would be different from the cloud. So I think you got confused that you should use "same" value for that env var (i.e. `PG_DATABASE_URL`) as that of cloud in local too. 

Does that make sense ? Feel free to point out any questions or concerns

**Reply by erquhart** [reply]
That was the point of confusion - I suspected this was the case, but the explanation around development mode cloning my cloud database gave me pause. I think I understand things now.

Opened this PR with a proposal to improve that section of the docs, feel free to close if it isn't helpful: https://github.com/hasura/graphql-engine/pull/8941

Thanks for your help!

**Reply by meetzaveri** [reply]
Glad that we were able to clear up the confusion, and thanks for raising PR.
